File: samples/akka-sample-cluster-java/src/main/java/sample/cluster/stats/StatsWorker.java
package sample.cluster.stats;

import java.time.Duration;
import java.util.HashMap;
import java.util.Map;

import akka.actor.typed.ActorRef;
import akka.actor.typed.Behavior;
import akka.actor.typed.javadsl.AbstractBehavior;
import akka.actor.typed.javadsl.ActorContext;
import akka.actor.typed.javadsl.Behaviors;
import akka.actor.typed.javadsl.Receive;
import sample.cluster.CborSerializable;

public final class StatsWorker extends AbstractBehavior<StatsWorker.Command> {

  interface Command extends CborSerializable {}
  public static final class Process implements Command {
    public final String word;
    public final ActorRef<Processed> replyTo;
    public Process(String word, ActorRef<Processed> replyTo) {
      this.word = word;
      this.replyTo = replyTo;
    }
  }
  private enum EvictCache implements Command {
    INSTANCE
  }
  public static final class Processed implements CborSerializable {
    public final String word;
    public final int length;
    public Processed(String word, int length) {
      this.word = word;
      this.length = length;
    }
  }

  private final Map<String, Integer> cache = new HashMap<String, Integer>();

  private StatsWorker(ActorContext<Command> context) {
    super(context);
  }

  public static Behavior<StatsWorker.Command> create() {
    return Behaviors.setup(context ->
        Behaviors.withTimers(timers -> {
          context.getLog().info("Worker starting up");
          timers.startTimerWithFixedDelay(EvictCache.INSTANCE, EvictCache.INSTANCE, Duration.ofSeconds(30));

          return new StatsWorker(context);
        })
    );
  }

  @Override
  public Receive<Command> createReceive() {
    return newReceiveBuilder()
        .onMessage(Process.class, this::process)
        .onMessageEquals(EvictCache.INSTANCE, this::evictCache)
        .build();
  }

  private Behavior<Command> evictCache() {
    cache.clear();
    return this;
  }

  private Behavior<Command> process(Process command) {
    getContext().getLog().info("Worker processing request [{}]", command.word);
    if (!cache.containsKey(command.word)) {
      int length = command.word.length();
      cache.put(command.word, length);
    }
    command.replyTo.tell(new Processed(command.word, cache.get(command.word)));
    return this;
  }
}

File: samples/akka-sample-cluster-scala/src/main/scala/sample/cluster/stats/StatsWorker.scala
package sample.cluster.stats

import akka.actor.typed.ActorRef
import akka.actor.typed.Behavior
import akka.actor.typed.scaladsl.ActorContext
import akka.actor.typed.scaladsl.Behaviors
import sample.cluster.CborSerializable

import scala.concurrent.duration._

//#worker
object StatsWorker {

  trait Command
  final case class Process(word: String, replyTo: ActorRef[Processed])
      extends Command
      with CborSerializable
  private case object EvictCache extends Command

  final case class Processed(word: String, length: Int) extends CborSerializable

  def apply(): Behavior[Command] = Behaviors.setup { ctx =>
    Behaviors.withTimers { timers =>
      ctx.log.info("Worker starting up")
      timers.startTimerWithFixedDelay(EvictCache, EvictCache, 30.seconds)

      withCache(ctx, Map.empty)
    }
  }

  private def withCache(ctx: ActorContext[Command],
                        cache: Map[String, Int]): Behavior[Command] =
    Behaviors.receiveMessage {
      case Process(word, replyTo) =>
        ctx.log.info("Worker processing request [{}]", word)
        cache.get(word) match {
          case Some(length) =>
            replyTo ! Processed(word, length)
            Behaviors.same
          case None =>
            val length = word.length
            val updatedCache = cache + (word -> length)
            replyTo ! Processed(word, length)
            withCache(ctx, updatedCache)
        }
      case EvictCache =>
        withCache(ctx, Map.empty)
    }
}
//#worker
